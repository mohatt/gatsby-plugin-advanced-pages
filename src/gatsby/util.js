import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import debug from 'debug'

// Initializes options
// The function only runs when called with args to initialize and cache
// options object. Subsequent calls without args return the cached value
export function initializeOptions (args) {
  if (arguments.length === 0) {
    if (typeof initializeOptions.options === 'undefined') {
      return reportError('Cant fetch options because they are not initialized')
    }

    return initializeOptions.options
  }

  // Merge user-defined options with defaults
  const options = _.merge(args.defaultOptions, args.pluginOptions)

  // set the root path of the the project
  options._root = args.store.getState().program.directory

  // Ensure basePath is absolute
  options.basePath = path.join('/', options.basePath)

  // Cache the options object now so that subsequent
  // calls dont throw an error
  initializeOptions.options = options

  // Verify directory locations
  options.directories = _.mapValues(options.directories, dir => {
    return lookupPath(dir)
  })

  // Verify default template
  if(options.template) {
    options.template = lookupPath(options.template, options.directories.templates)
  }

  // Debug final options object
  debug('gatsby-plugin-advanced-pages')('Options', options)

  initializeOptions.options = options
}

// Initializes reporter
// The function only runs when called with args to initialize and cache
// reporter object. Subsequent calls without args return the cached value
export function initializeReporter (reporter) {
  if (arguments.length === 0) {
    if (typeof initializeReporter.reporter === 'undefined') {
      throw new Error('Cant fetch reporter because it is not initialized')
    }

    return initializeReporter.reporter
  }

  const errorMap = {
    [10000]: {
      text: context => context.message,
      level: `ERROR`,
      type: `PLUGIN`,
    },
  }

  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }

  initializeReporter.reporter = reporter
}

// Gets the initialized eporter object
export function getReporter () {
  return initializeReporter()
}

// Gets the initialized eporter object
export function reportError (message, e = null) {
  const reporter = getReporter()
  const prefix = `"gatsby-plugin-advanced-pages" threw an error while running`

  if(!e) {
    reporter.panic({
      id: `10000`,
      context: {
        message: prefix
      },
      error: new Error(message)
    })
    return
  }

  reporter.panic({
    id: `10000`,
    context: {
      message: `${prefix}:\n ${message}`
    },
    error: e
  })
}

// Gets the initialized options object
export function getOptions () {
  return initializeOptions()
}

// Gets the value of a single option
export function getOption (optionName) {
  return _.get(getOptions(), optionName)
}

// Checks if a location exists under parent
// if not checks if it exists under project root
// if not check if its an absolute path
// throws an error if cant find any
export function lookupPath(location, parent = null){
  if(!location || typeof location !== 'string') {
    return reportError(
      `ensurePath() expected a non-empty string but got ${typeof location}("${location}")`
    )
  }

  let search = []
  if(parent) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if(fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(getOption('_root'), location)
  search.push(rootPath)
  if(fs.existsSync(rootPath)) {
    return rootPath
  }

  search.push(location)
  if(fs.existsSync(location)) {
    return location
  }

  reportError(
    `A path with value "${location}" could not be found at ` +
    `any of the following locations:\n - "${search.join(`\n - "`)}"`
  )
}
