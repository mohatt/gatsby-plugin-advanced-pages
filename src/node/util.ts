import fs from 'fs'
import path from 'path'
import { createRequire } from 'module'
import Yaml from 'js-yaml'
import { validateOptionsSchema, Joi, PluginOptionsSchemaJoi, Schema } from 'gatsby-plugin-utils'
import _ from 'lodash'
import debug from 'debug'
import type { GatsbyNode, NodePluginArgs, Reporter } from 'gatsby'
import type { DefaultPluginOptions, PluginOptions } from './api'

/**
 * Represents the resolved internal options for the plugin.
 * It includes additional properties required for internal processing.
 */
interface ResolvedPluginOptions extends DefaultPluginOptions {
  /**
   * Absolute path to the root of the Gatsby project.
   * Used to resolve relative paths for templates, directories, etc.
   */
  _root: string
}

/**
 * Provides functionality to initialize, merge, and retrieve plugin options.
 *
 * @internal
 */
export const options = (() => {
  let ref: ResolvedPluginOptions | undefined

  /**
   * Initializes and merges plugin options with defaults.
   */
  const initialize = (args: {
    store: NodePluginArgs['store']
    pluginOptions: PluginOptions
    defaultOptions: DefaultPluginOptions
  }): ResolvedPluginOptions => {
    // Merge user-defined options with defaults
    const opts: ResolvedPluginOptions = _.merge(
      // set the root path of the project
      { _root: args.store.getState().program.directory },
      args.defaultOptions,
      args.pluginOptions,
    )
    // Normalize options
    Object.assign(opts, {
      // Ensure basePath is absolute
      basePath: path.join('/', opts.basePath),
      // Ensure type names are Title case
      typeNames: _.mapValues(opts.typeNames, _.upperFirst),
      // Convert directory paths to absolute ones
      directories: _.mapValues(opts.directories, (dir) => path.join(opts._root, dir)),
    })

    // Cache the options object now so that subsequent
    // calls to ensurePath() don't throw an error
    ref = opts

    // Verify default template exists
    if (opts.template != null) {
      opts.template = ensurePath(opts.template, opts.directories.templates)
    }

    // Debug final options object
    debug('gatsby-plugin-advanced-pages')('Options', opts)

    ref = opts

    return opts
  }

  /**
   * Retrieves a specific option value or the entire options object.
   *
   * - If called without arguments, returns the entire options object.
   * - If called with a key, retrieves the corresponding value.
   * - Throws an error if options are not initialized.
   *
   * @param optionName (Optional) The key of the option to retrieve.
   *
   * @returns The requested option value or the entire options object.
   *
   * @throws - Terminates execution if options are not initialized.
   */
  const get: {
    (): ResolvedPluginOptions
    <T extends keyof ResolvedPluginOptions>(optionName: T): ResolvedPluginOptions[T]
  } = (...args: [] | [string]) => {
    if (ref === undefined) {
      // throw new Error('Plugin options are not initialized.')
      throw new PluginError('Cant fetch options because they are not initialized')
    }

    if (args.length === 0) {
      return ref
    }

    return _.get(ref, ...args)
  }

  return { initialize, get }
})()

/**
 * Manages Gatsby's reporter instance, providing functionalities for structured error logging,
 * warnings, and instance management.
 *
 * @internal
 */
export const reporter = (() => {
  let ref: Reporter | undefined

  // Initializes reporter
  /**
   * Initializes and caches the Gatsby reporter instance.
   *
   * If `setErrorMap` is available, it registers custom error messages
   * for better debugging in Gatsby's CLI.
   */
  const initialize = (instance: Reporter): Reporter => {
    if (instance.setErrorMap != null) {
      instance.setErrorMap({
        10000: {
          text: (context) => context.message,
          category: 'USER',
          level: 'ERROR',
          type: 'PLUGIN',
        },
      })
    }

    ref = instance
    return instance
  }

  /**
   * Retrieves the cached reporter instance.

   * @throws Error if the reporter is accessed before being initialized.
   */
  const get = (): Reporter => {
    if (ref === undefined) {
      throw new Error('Cant fetch reporter instance because it is not initialized')
    }

    return ref
  }

  /**
   * Logs an error and terminates the build process.
   *
   * - Uses Gatsby's `panic()` method to log structured errors.
   * - If an `Error` object is provided, includes it in the error report.
   * - Otherwise, creates a new error instance from the given message.
   *
   * @param api The Gatsby Node API that caused the error.
   *
   * @param titleOrError The error message or PluginError object to display.
   *
   * @param error (Optional) The original error instance.
   *
   * @throws - This function never returns; it always terminates execution.
   */
  const error = (
    api: keyof GatsbyNode,
    titleOrError: string | PluginError | Error,
    error?: Error,
  ): never => {
    const instance = get()
    const prefix = `"gatsby-plugin-advanced-pages" threw an error during "${api}" hook`

    let title: string | undefined
    let mainError: Error | undefined

    if (titleOrError instanceof PluginError) {
      title = titleOrError.message
      mainError = titleOrError.originalError
    } else if (titleOrError instanceof Error) {
      mainError = titleOrError
    } else {
      title = titleOrError
    }

    if (error) {
      mainError = error
    }

    return instance.panic({
      id: '10000',
      context: {
        message: mainError && title ? `${prefix}:\n ${title}` : prefix,
      },
      error: mainError ?? (title ? new Error(title) : undefined),
    })
  }

  /**
   * Logs a warning message in the Gatsby CLI.
   *
   * @param message The warning message to display.
   */
  const warning = (message: string): void => {
    const instance = get()
    const prefix = '"gatsby-plugin-advanced-pages" might not be working properly'
    instance.warn(`${prefix}:\n ${message}`)
  }

  return { initialize, get, error, warning }
})()

/**
 * Validates the input data using Joi schema validation.
 *
 * @param schema The Joi schema (or schema factory) to validate against.
 * @param data The input data to validate.
 * @param [name] (Optional) The prop name of the data being validated. Default is `data`.
 *
 * @returns A promise that resolves to the validated and typed data.
 *
 * @throws {ValidationError} Throws an error if the input does not meet the schema requirements.
 */
export const validateSchema = async <E>(
  schema: Schema | ((Joi: PluginOptionsSchemaJoi) => Schema),
  data: unknown,
  name = 'data',
): Promise<E> => {
  const objSchema = Joi.object({ [name]: schema instanceof Function ? schema(Joi) : schema })
  const { value } = await validateOptionsSchema(objSchema, { [name]: data })
  return value[name] as E
}

/**
 * Ensures a path exists based on different search locations:
 *
 * The function checks in the following order:
 * 1. If `location` exists under the given `parent` directory, return it.
 * 2. If not found, checks under the Gatsby project root (`_root`).
 * 3. If still not found and `location` is absolute, it verifies the path exists.
 * 4. If none of the above match, logs an error and terminates execution.
 *
 * @param location The relative or absolute file path to resolve.
 * @param parent (Optional) A parent directory to check first.
 *
 * @returns The absolute path if found; otherwise, throws an error.
 *
 * @throws PluginError if the path is not found.
 */
export const ensurePath = (location: string, parent?: string): string => {
  const search = []
  if (parent != null) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if (fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(options.get('_root'), location)
  search.push(rootPath)
  if (fs.existsSync(rootPath)) {
    return rootPath
  }

  if (path.isAbsolute(location)) {
    search.push(location)
    if (fs.existsSync(location)) {
      return location
    }
  }

  throw new PluginError(
    `A path with value "${location}" could not be found at ` +
      `any of the following locations:\n - "${search.join('\n - "')}"`,
  )
}

/**
 * Retrieves the default export of a module.
 */
export const getDefaultExport = async <T = unknown>(modulePath: string): Promise<T> => {
  const noop = {}
  let defaultExport: any = noop

  // Avoid require in tests since Vitest doesn't support mocking require calls
  if (process.env.VITEST !== 'true') {
    const nodeRequire = require ?? createRequire(import.meta.url)
    try {
      defaultExport = nodeRequire(modulePath)
      if (defaultExport?.__esModule === true && 'default' in defaultExport) {
        defaultExport = defaultExport.default
      }
    } catch (error) {
      if (error.code !== 'ERR_REQUIRE_ESM') {
        throw new PluginError(`Unable to import module "${modulePath}"`, error)
      }
    }
  }

  if (defaultExport === noop) {
    try {
      const esmExports = await import(modulePath)
      defaultExport = esmExports.default
    } catch (error) {
      throw new PluginError(`Unable to import ESM module "${modulePath}"`, error)
    }
  }

  if (defaultExport == null) {
    throw new PluginError(
      `Unable to import module "${modulePath}"`,
      new Error(`Couldn't find valid default export`),
    )
  }

  return defaultExport as T
}

/**
 * Searches for a configuration file under the given root directory and loads its content.
 *
 * @param root The root directory to search in.
 * @param baseName The base name of the config file (e.g., `"pages.config"`).
 *
 * @returns An object containing the file path and its parsed data, or `null` if not found.
 */
export const findConfigFile = async (
  root: string,
  baseName: string,
): Promise<{ path: string; data: unknown } | null> => {
  const extensions = ['yaml', 'yml', 'json', 'js', 'cjs', 'mjs']

  for (const ext of extensions) {
    const filePath = path.join(root, `${baseName}.${ext}`)
    if (!fs.existsSync(filePath)) {
      continue
    }

    try {
      switch (ext) {
        case 'yaml':
        case 'yml':
          return {
            path: filePath,
            data: Yaml.load(fs.readFileSync(filePath, 'utf8'), {
              filename: filePath,
              schema: Yaml.FAILSAFE_SCHEMA,
            }),
          }

        case 'json':
          return {
            path: filePath,
            data: JSON.parse(fs.readFileSync(filePath, 'utf8')),
          }

        default:
          return { path: filePath, data: await getDefaultExport(filePath) }
      }
    } catch (error) {
      throw new PluginError(`Failed to load config file "${filePath}"`, error)
    }
  }

  return null
}

/**
 * Wraps a Gatsby API function with error handling.
 *
 * @param api - The name of the Gatsby API being wrapped (e.g., `"onCreatePage"`).
 * @param fn - The original Gatsby API function to be wrapped.
 *
 * @returns The wrapped function with error handling.
 */
export const createPluginExport = <T extends keyof GatsbyNode, F extends GatsbyNode[T]>(
  api: T,
  fn: F,
): F => {
  return ((...args: any[]) => {
    try {
      // Call the original function with provided arguments
      const result = (fn as any)(...args)

      // If the function returns a Promise, attach error handling
      if (result instanceof Promise) {
        return result.catch((error) => reporter.error(api, error))
      }

      // Return the result (for synchronous functions)
      return result
    } catch (error) {
      // Catch synchronous errors and report them
      reporter.error(api, error)
    }
  }) as F
}

/**
 * Custom error class for plugin-related errors.
 */
export class PluginError extends Error {
  /**
   * @param message - A description of the error.
   * @param originalError - The original error that caused this issue (optional).
   */
  constructor(
    message: string,
    public readonly originalError?: Error,
  ) {
    super(message)
  }
}
