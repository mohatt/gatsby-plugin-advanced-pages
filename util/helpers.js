const path = require('path')
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Checks if a path is a child of another one
function isPathChildof (child, parent) {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  isPathChildof,
}
