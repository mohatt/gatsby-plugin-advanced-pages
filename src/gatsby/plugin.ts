import type { GatsbyNode, PluginOptions as GatsbyPluginOptions } from 'gatsby'
import { options, reporter } from './util'
import { getOptionsSchema } from './schema'
import type { PluginOptions } from './index'

// Validates user-defined options against schema
// runs before onPreBootstrap
export const pluginOptionsSchema: GatsbyNode['pluginOptionsSchema'] = ({ Joi }) => {
  return getOptionsSchema(Joi)
}

// Initializes plugin state
export const onPluginInit: GatsbyNode['onPluginInit'] = (
  args,
  pluginOptions: GatsbyPluginOptions & PluginOptions,
) => {
  // Default values for options
  const defaultOptions: PluginOptions = {
    basePath: '/',
    pages: [],
    template: null,
    directories: {
      templates: './src/templates',
      helpers: './gatsby/pages',
    },
    pagination: {
      limit: 10,
      suffix: '/page/:page',
    },
    typeNames: {
      page: 'Page',
      pageRoute: 'PageRoute',
    },
  }

  // Initializes and validates options
  // Only runs once at bootstrap
  options.initialize({
    store: args.store,
    defaultOptions,
    pluginOptions,
  })

  // Initializes plugin reporter
  reporter.initialize(args.reporter)
}

// Creates a webpack alias for the plugin cache directory
// Usually located at <root>/.cache/caches/<plugin>
export const onCreateWebpackConfig: GatsbyNode['onCreateWebpackConfig'] = ({ actions, cache }) => {
  actions.setWebpackConfig({
    resolve: {
      alias: {
        'gatsby-plugin-advanced-pages-cache': cache.directory,
      },
    },
  })
}

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = ({
  actions,
  schema,
}) => {
  const { createTypes } = actions
  const { page, pageRoute } = options.get('typeNames')

  createTypes([
    schema.buildObjectType({
      name: pageRoute,
      fields: {
        name: 'String!',
        path: 'String!',
      },
    }),
    schema.buildObjectType({
      name: page,
      fields: {
        title: 'String!',
        templateName: 'String',
        template: 'String',
        helper: 'String',
        data: 'JSON',
        routes: `[${pageRoute}!]!`,
      },
      interfaces: ['Node'],
    }),
  ])
}

export { default as sourceNodes } from './source-nodes'
export { default as createPages } from './create-pages'
