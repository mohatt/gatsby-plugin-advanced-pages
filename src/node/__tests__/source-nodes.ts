import { mountFile, mountModule, setupPlugin, resetVFS } from '@test/util'
import testCases from './__fixtures__/source-nodes'
import { options as pluginOptions } from '../util'
import { sourceNodes } from '../plugin'

describe('sourceNodes', () => {
  const createNode = vi.fn()
  const createNodeId = vi.fn().mockImplementation((v) => ('id-' + v).toUpperCase())
  const createContentDigest = vi.fn().mockImplementation((v) => JSON.stringify(v))
  const actions = { createNode }
  const args: any = {
    actions,
    createNodeId,
    createContentDigest,
  }

  beforeEach(() => {
    setupPlugin()
    resetVFS()
    createNode.mockReset()
  })

  it.each(testCases)('$title', async ({ files, throws, options }) => {
    if (options) setupPlugin(options)
    files.forEach((file) => {
      if (typeof file === 'string') mountFile(file)
      else if (Array.isArray(file.data)) mountModule(file.path, file.data)
      else mountFile(file.path, file.data)
    })

    const promise = expect(sourceNodes(args))
    if (throws) {
      await promise.rejects.toMatchSnapshot()
      return
    }

    await promise.resolves.toBeUndefined()
    expect(createNode.mock.calls).toMatchSnapshot()
    expect(pluginOptions.get('_pages')).toMatchSnapshot()
  })
})
