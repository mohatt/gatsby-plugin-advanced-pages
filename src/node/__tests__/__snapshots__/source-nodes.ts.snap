// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sourceNodes > 'correctly creates multiple page nodes with various configurations' 1`] = `
[
  [
    {
      "children": [],
      "data": {},
      "helper": "/virtual/project/gatsby/pages/helper.js",
      "id": "ID-0 - FOO",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO","title":"foo","template":"/virtual/project/src/templates/default.js","helper":"/virtual/project/gatsby/pages/helper.js","routes":[{"name":"foo","path":"/foo"},{"name":"foo.bar","path":"/foo/bar"}],"data":{},"templateName":"default.js"}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo",
        },
        {
          "name": "foo.bar",
          "path": "/foo/bar",
        },
      ],
      "template": "/virtual/project/src/templates/default.js",
      "templateName": "default.js",
      "title": "foo",
    },
  ],
  [
    {
      "children": [],
      "data": {},
      "helper": undefined,
      "id": "ID-1 - LOREM",
      "internal": {
        "contentDigest": "{"id":"ID-1 - LOREM","title":"lorem","template":"/virtual/project/src/templates/test.js","routes":[{"name":"lorem","path":"/lorem"}],"data":{},"templateName":"test.js"}",
        "description": "Advanced Page: lorem",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "lorem",
          "path": "/lorem",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "test.js",
      "title": "lorem",
    },
  ],
  [
    {
      "children": [],
      "data": {
        "name": "john",
        "title": "don",
      },
      "helper": "/virtual/project/gatsby/custom/helper.js",
      "id": "ID-2 - IPSUM",
      "internal": {
        "contentDigest": "{"id":"ID-2 - IPSUM","title":"ipsum","template":"/virtual/project/src/custom/template.js","helper":"/virtual/project/gatsby/custom/helper.js","data":{"name":"john","title":"don"},"routes":[{"name":"ipsum","path":"/ipsum"},{"name":"dolor","path":"/dolor"},{"name":"sit","path":"/sit/amet"}],"templateName":"./src/custom/template.js"}",
        "description": "Advanced Page: ipsum",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "ipsum",
          "path": "/ipsum",
        },
        {
          "name": "dolor",
          "path": "/dolor",
        },
        {
          "name": "sit",
          "path": "/sit/amet",
        },
      ],
      "template": "/virtual/project/src/custom/template.js",
      "templateName": "./src/custom/template.js",
      "title": "ipsum",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly creates multiple page nodes with various configurations' 2`] = `
{
  "ID-0 - FOO": {
    "data": {},
    "helper": "helper.js",
    "routes": {
      "foo": "/foo",
      "foo.bar": "/foo/bar",
    },
    "template": "default.js",
    "title": "foo",
  },
  "ID-1 - LOREM": {
    "data": {},
    "routes": {
      "lorem": "/lorem",
    },
    "template": "test.js",
    "title": "lorem",
  },
  "ID-2 - IPSUM": {
    "data": {
      "name": "john",
      "title": "don",
    },
    "helper": "./gatsby/custom/helper.js",
    "routes": {
      "dolor": "/dolor",
      "ipsum": "/ipsum",
      "sit": "/sit/amet",
    },
    "template": "./src/custom/template.js",
    "title": "ipsum",
  },
}
`;

exports[`sourceNodes > 'correctly creates page nodes from inline pages config' 1`] = `
[
  [
    {
      "children": [],
      "helper": undefined,
      "id": "ID-0 - FOO",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO","title":"foo","template":"/virtual/project/src/templates/test.js","routes":[{"name":"foo","path":"/foo"}],"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
  [
    {
      "children": [],
      "helper": "/virtual/project/gatsby/custom/helper.js",
      "id": "ID-1 - BAR-FILE",
      "internal": {
        "contentDigest": "{"id":"ID-1 - BAR-FILE","title":"bar-file","template":"/virtual/project/src/templates/test.js","helper":"/virtual/project/gatsby/custom/helper.js","routes":[{"name":"bar1","path":"/bar-file"}],"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: bar-file",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "bar1",
          "path": "/bar-file",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "bar-file",
    },
  ],
  [
    {
      "children": [],
      "helper": "[Function]",
      "id": "ID-2 - BAR-INLINE",
      "internal": {
        "contentDigest": "{"id":"ID-2 - BAR-INLINE","title":"bar-inline","template":"/virtual/project/src/templates/test.js","helper":"[Function]","routes":[{"name":"bar2","path":"/bar-inline"}],"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: bar-inline",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "bar2",
          "path": "/bar-inline",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "bar-inline",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly creates page nodes from inline pages config' 2`] = `
{
  "ID-0 - FOO": {
    "routes": {
      "foo": "/foo",
    },
    "template": "./src/templates/test.js",
    "title": "foo",
  },
  "ID-1 - BAR-FILE": {
    "helper": "./gatsby/custom/helper.js",
    "routes": {
      "bar1": "/bar-file",
    },
    "template": "./src/templates/test.js",
    "title": "bar-file",
  },
  "ID-2 - BAR-INLINE": {
    "helper": [Function],
    "routes": {
      "bar2": "/bar-inline",
    },
    "template": "./src/templates/test.js",
    "title": "bar-inline",
  },
}
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.js' 1`] = `
[
  [
    {
      "children": [],
      "data": {},
      "helper": undefined,
      "id": "ID-0 - FOO",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO","title":"foo","template":"/virtual/project/src/templates/test.js","routes":[{"name":"foo","path":"/foo"}],"data":{},"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
  [
    {
      "children": [],
      "data": {},
      "helper": "/virtual/project/gatsby/pages/test.js",
      "id": "ID-1 - BAR",
      "internal": {
        "contentDigest": "{"id":"ID-1 - BAR","title":"bar","template":"/virtual/project/src/templates/test.js","helper":"/virtual/project/gatsby/pages/test.js","routes":[{"name":"bar","path":"/bar"}],"data":{},"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: bar",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "bar",
          "path": "/bar",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "bar",
    },
  ],
  [
    {
      "children": [],
      "data": {},
      "helper": "[Function]",
      "id": "ID-2 - BAR-INLINE",
      "internal": {
        "contentDigest": "{"id":"ID-2 - BAR-INLINE","title":"bar-inline","template":"/virtual/project/src/templates/test.js","helper":"[Function]","routes":[{"name":"bar2","path":"/bar-inline"}],"data":{},"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: bar-inline",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "bar2",
          "path": "/bar-inline",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "bar-inline",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.js' 2`] = `
{
  "ID-0 - FOO": {
    "data": {},
    "routes": {
      "foo": "/foo",
    },
    "template": "./src/templates/test.js",
    "title": "foo",
  },
  "ID-1 - BAR": {
    "data": {},
    "helper": "test.js",
    "routes": {
      "bar": "/bar",
    },
    "template": "./src/templates/test.js",
    "title": "bar",
  },
  "ID-2 - BAR-INLINE": {
    "data": {},
    "helper": [Function],
    "routes": {
      "bar2": "/bar-inline",
    },
    "template": "./src/templates/test.js",
    "title": "bar-inline",
  },
}
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.json' 1`] = `
[
  [
    {
      "children": [],
      "data": {},
      "helper": undefined,
      "id": "ID-0 - FOO",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO","title":"foo","template":"/virtual/project/src/templates/test.js","routes":[{"name":"foo","path":"/foo"}],"data":{},"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.json' 2`] = `
{
  "ID-0 - FOO": {
    "data": {},
    "routes": {
      "foo": "/foo",
    },
    "template": "./src/templates/test.js",
    "title": "foo",
  },
}
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.yaml' 1`] = `
[
  [
    {
      "children": [],
      "data": {},
      "helper": undefined,
      "id": "ID-0 - FOO",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO","title":"foo","template":"/virtual/project/src/templates/test.js","routes":[{"name":"foo","path":"/foo"}],"data":{},"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly creates page nodes from pages.config.yaml' 2`] = `
{
  "ID-0 - FOO": {
    "data": {},
    "routes": {
      "foo": "/foo",
    },
    "template": "./src/templates/test.js",
    "title": "foo",
  },
}
`;

exports[`sourceNodes > 'correctly ignores pages config file in favor of inline pages config' 1`] = `
[
  [
    {
      "children": [],
      "helper": undefined,
      "id": "ID-0 - FOO-INLINE",
      "internal": {
        "contentDigest": "{"id":"ID-0 - FOO-INLINE","title":"foo-inline","template":"/virtual/project/src/templates/test.js","routes":[{"name":"foo","path":"/foo-inline"}],"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: foo-inline",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "foo",
          "path": "/foo-inline",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo-inline",
    },
  ],
  [
    {
      "children": [],
      "helper": "[Function]",
      "id": "ID-1 - BAR-INLINE",
      "internal": {
        "contentDigest": "{"id":"ID-1 - BAR-INLINE","title":"bar-inline","template":"/virtual/project/src/templates/test.js","helper":"[Function]","routes":[{"name":"bar","path":"/bar-inline"}],"templateName":"./src/templates/test.js"}",
        "description": "Advanced Page: bar-inline",
        "type": "Page",
      },
      "parent": null,
      "routes": [
        {
          "name": "bar",
          "path": "/bar-inline",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "bar-inline",
    },
  ],
]
`;

exports[`sourceNodes > 'correctly ignores pages config file in favor of inline pages config' 2`] = `
{
  "ID-0 - FOO-INLINE": {
    "routes": {
      "foo": "/foo-inline",
    },
    "template": "./src/templates/test.js",
    "title": "foo-inline",
  },
  "ID-1 - BAR-INLINE": {
    "helper": [Function],
    "routes": {
      "bar": "/bar-inline",
    },
    "template": "./src/templates/test.js",
    "title": "bar-inline",
  },
}
`;

exports[`sourceNodes > 'does not throw for an empty pages config file' 1`] = `[]`;

exports[`sourceNodes > 'does not throw for an empty pages config file' 2`] = `{}`;

exports[`sourceNodes > 'shows a warning for non existing pages config file' 1`] = `
{
  "error": [Warning: "gatsby-plugin-advanced-pages" might not be working properly:
 - No pages config is defined in plugin options.
 - Unable to find a valid pages config file (eg. "pages.config.js") under "/virtual/project"],
  "text": "The plugin threw an error during "sourceNodes" hook",
}
`;

exports[`sourceNodes > 'throws for invalid JSON syntax in pages.config.json' 1`] = `
{
  "error": [SyntaxError: Unexpected end of JSON input],
  "text": "The plugin threw an error during "sourceNodes" hook:
 Failed to load config file "/virtual/project/pages.config.json"",
}
`;

exports[`sourceNodes > 'throws for invalid YAML syntax in pages.config.yaml' 1`] = `
{
  "error": [YAMLException: can not read an implicit mapping pair; a colon is missed in "/virtual/project/pages.config.yaml" at line 7, column 21:
      invalid_yaml_entry
                        ^],
  "text": "The plugin threw an error during "sourceNodes" hook:
 Failed to load config file "/virtual/project/pages.config.yaml"",
}
`;

exports[`sourceNodes > 'throws for invalid pages config file schema' 1`] = `
{
  "error": [ValidationError: "pages[0].routes.foo": Route paths must be strings that start with '/'],
  "text": "The plugin threw an error during "sourceNodes" hook:
 Unable to validate pages config file "/virtual/project/pages.config.yaml"",
}
`;

exports[`sourceNodes > 'throws for pages with invalid template' 1`] = `
{
  "error": [Error: A path with value "./no/template.js" could not be found at any of the following locations:
 - "/virtual/project/src/templates/no/template.js
 - "/virtual/project/no/template.js"],
  "text": "The plugin threw an error during "sourceNodes" hook",
}
`;

exports[`sourceNodes > 'throws for pages with no template' 1`] = `
{
  "error": [Error: Missing "template" metadata for page[0]: "foo". No default template is set in plugin options either.],
  "text": "The plugin threw an error during "sourceNodes" hook",
}
`;
